lib/init.c

```diff
@@ -11,24 +11,26 @@ int metal_init(const struct metal_init_params *params)
 {
 	int error = 0;
 
-	memset(&_metal, 0, sizeof(_metal));
+	if (atomic_fetch_add(&_metal.common.ref_count, 1) == 0) {
+		_metal.common.log_handler = params->log_handler;
+		_metal.common.log_level   = params->log_level;
 
-	_metal.common.log_handler   = params->log_handler;
-	_metal.common.log_level     = params->log_level;
+		metal_list_init(&_metal.common.bus_list);
+		metal_list_init(&_metal.common.generic_shmem_list);
+		metal_list_init(&_metal.common.generic_device_list);
 
-	metal_list_init(&_metal.common.bus_list);
-	metal_list_init(&_metal.common.generic_shmem_list);
-	metal_list_init(&_metal.common.generic_device_list);
-
-	error = metal_sys_init(params);
-	if (error)
-		return error;
+		error = metal_sys_init(params);
+		if (error) {
+			atomic_fetch_sub(&_metal.common.ref_count, 1);
+			return error;
```

Comment:
The return was and is still useless . could be suppressed

Reply:
Do you want to change the return value of metal_init and metal_sys_init to void or just ignore the value returned from metal_sys_init?

Reply:
unclear..., it was just to suppress the "return error;" duplication
My proposal was to have something like this
```
int metal_init(const struct metal_init_params *params)
{
	int error = 0;
	if (atomic_fetch_add(&_metal.common.ref_count, 1) == 0) {

[...]
		error = metal_sys_init(params);
		if (error) {
			atomic_fetch_sub(&_metal.common.ref_count, 1);
}

	return error;
```

Reply:
Done.